### PR TITLE
Add refactorings from fragments branch

### PR DESCRIPTION
**Pull request by daun**
The fragments feature won't be part of swup 3, but the feature branch has a few things we'd like to merge already:

- General refactoring, simplifications and formatting
- Add early returns to event handlers to avoid excessive nesting of if/else blocks
- Extract `leavePage` and `enterPage` into separate modules/steps, making work on future features easier
- Add `nextTick` helper for animation timings
- Add `updateHistoryRecord` helper in addition to existing `createHistoryRecord`

Tests are still passing.

### CODE REVIEW

File: src/helpers/createHistoryRecord.js [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
-		{
-			url: url || window.location.href.split(window.location.hostname)[1],
-			random: Math.random(),
-			source: 'swup'
-		},
-		document.title,
-		url || window.location.href.split(window.location.hostname)[1]
-	);
+const createHistoryRecord = (url, customData = {}) => {
+	url = url || window.location.href.split(window.location.hostname)[1];
```
Comment (hirasso) [inline src/helpers/createHistoryRecord.js:2]: While we are at it, couldn't we use swups `getCurrentUrl()` here?
```js
import getCurrentUrl from "./getCurrentUrl";
const createHistoryRecord = (url, customData = {}) => {
	url = url || getCurrentUrl();
    // ...
}
```

File: src/helpers/updateHistoryRecord.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const updateHistoryRecord = (url = null, customData = {}) => {
+	url = url || window.location.href; // window.location.href.split(window.location.hostname)[1];
```
Comment (hirasso) [inline src/helpers/updateHistoryRecord.js:2]: Same as with `createHistoryRecord`. Or what was the reason the full URL is being used here?

File: src/helpers/createHistoryRecord.js (hunk 1 above)
Comment (hirasso) [reply src/helpers/createHistoryRecord.js:2]: Also, I always found `getCurrentUrl` always a bit confusing, naming-wise. How about we change it to `getCurrentPath` and add an alias for backwards-compatibility, what do you think?

File: src/index.js [hunk 3]
```diff
@@ -196,65 +201,57 @@ export default class Swup {
...
+				if (element) {
+					updateHistoryRecord(url + hash);
+				} else {
+					console.warn(`Element for offset not found (#${hash})`);
 				}
 			}
-		} else {
-			// open in new tab (do nothing)
-			this.triggerEvent('openPageInNewTab', event);
+		} else if (!this.isSameResolvedPath(url, getCurrentUrl())) {
```
Comment (hirasso) [inline src/index.js:248]: Here would be another chance to do an early returns. If you agree, I could push the change directly to this branch.

File: src/modules/loadPage.js [hunk 4]
```diff
@@ -1,100 +1,66 @@
...
 
 	this.triggerEvent('transitionStart', popstate);
 
 	// set transition object
-	if (data.customTransition != null) {
-		this.updateTransition(window.location.pathname, data.url, data.customTransition);
-		document.documentElement.classList.add(`to-${classify(data.customTransition)}`);
-	} else {
-		this.updateTransition(window.location.pathname, data.url);
+	this.updateTransition(window.location.pathname, url, customTransition);
```
Comment (hirasso) [inline src/modules/loadPage.js:13]: I know, not related to this PR, but shouldn't we use `getCurrentUrl` here as well?
```js
this.updateTransition(getCurrentUrl(), url, customTransition);
```

**Review by hirasso — COMMENTED** [review]
Very nice!! 🍰 Added a few comments/questions.

File: src/helpers/createHistoryRecord.js (hunk 1 above)
Comment (daun) [reply src/helpers/createHistoryRecord.js]: Done :) I'd prefer not to touch the naming, just maintenance-wise.

File: src/helpers/updateHistoryRecord.js (hunk 2 above)
Comment (daun) [reply src/helpers/updateHistoryRecord.js]: Done :) It was just inconsistent since it doesn't matter in this context if the URL is relative or absolute.

File: src/index.js (hunk 3 above)
Comment (daun) [reply src/index.js:240]: This one I'm unsure about. Do we exit early here before the first `if` check or after? Maybe you tackle that one.

File: src/modules/loadPage.js (hunk 4 above)
Comment (daun) [reply src/modules/loadPage.js]: Done :) It has the added benefit that now the query params end up in the transition object as well.

File: src/index.js (hunk 3 above)
Comment (hirasso) [reply src/index.js:240]: Ok, I'll take care of this.

Comment (hirasso) [reply src/index.js]: Did some more refactoring. Looking forward to hear what you think! Tests are passing. I also ran `lint` on the code, which turned the option `ignoreLink` to this:

```js
let defaults = {
	//...
	ignoreLink: (el) => 
		el.origin !== window.location.origin || el.closest('[data-no-swup]'),
	//...
};
```
It becomes slightly more readable by an explicit return IMHO:
```js
let defaults = {
	// ...
	ignoreLink: (el) => {
		return el.origin !== window.location.origin || el.closest('[data-no-swup]');
	},
	// ...
};
```

File: src/helpers/createHistoryRecord.js (hunk 1 above)
Comment (hirasso) [reply src/helpers/createHistoryRecord.js]: Hmm. I really find it confusing that we are using `getCurrentUrl()`, but then we also use `resolvePath` on it. We should probably either 

1. rename `resolvePath()` to `resolveUrl()`
2. or rename `getCurrentUrl()` to  `getCurrentPath()`

Since the thing actually _is_ a relative URL composed by `pathname` + `searchParams`, and also for backwards-compatibility, I would tend to rename the new option #1 then. What do you think?

File: src/index.js (hunk 3 above)
Comment (daun) [reply src/index.js]: The reason I left it unlinted before is for readability. As it is, the one-liner works fine, but once you add more checks, the previous one-check-per-line-format is a lot more readable in my opinion, especially when explaining it in the docs. How do you feel about reverting it and adding a `prettier-ignore` above? Too fussy?

File: src/helpers/createHistoryRecord.js (hunk 1 above)
Comment (daun) [reply src/helpers/createHistoryRecord.js]: In that case, I also prefer option 1, renaming the new option `resolveUrl`. Could you do that for the main repo and the docs? And the scroll plugin, I guess, since that's already using it.

File: src/index.js (hunk 3 above)
Comment (hirasso) [reply src/index.js]: > The reason I left it unlinted before is for readability. As it is, the one-liner works fine, but once you add more checks, the previous one-check-per-line-format is a lot more readable in my opinion, especially when explaining it in the docs. How do you feel about reverting it and adding a `prettier-ignore` above? Too fussy?

Hmm, not a fan of adding prettier-ignore there. I would then maybe prefer walking each condition with guard clauses. This will make it VERY clear what's happening. We could even add inline documentation:
```js
let defaults = {
	// ...
	ignoreLink: (el) => {
		// ignore external links:
		if (el.origin !== window.location.origin) return true;
		// ignore links that are inside a container with the attribute [data-no-swup]:
		if (el.closest('[data-no-swup]')) return true;
		// handle all other links
		return false;
	},
	// ...
};
```

**Comment by daun** [comment]
@hirasso Ready to merge? 🧃

File: src/index.js (hunk 3 above)
Comment (hirasso) [reply src/index.js]: Yes we shall! :)

Comment (hirasso) [reply src/index.js]: ...ready to merge then!

Comment (daun) [reply src/index.js]: 🚀

File: src/helpers/updateHistoryRecord.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { getCurrentUrl } from '../helpers.js';
+
+const updateHistoryRecord = (url = null, customData = {}) => {
+	url = url || getCurrentUrl({ hash: true });
```
Comment (hirasso) [inline src/helpers/updateHistoryRecord.js:4]: Why did you add the optional hash here? Missed that before

Comment (hirasso) [reply src/helpers/updateHistoryRecord.js:4]: Sorry, wrong context... why does `getCurrentUrl()` now accept returning the hash as well?

Comment (daun) [reply src/helpers/updateHistoryRecord.js:4]: `getCurrentUrl` by default returns the url without a hash, but updating the history record should keep any current hash. I figured it's clearner to make it an option in `getCurrentUrl` than to combine `getCurrentUrl() + location.hash` or something. Should we rethink that? Not too late 🤡
